compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -323,13 +323,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 let mut param_found = FxHashMap::default();
                 if self.can_eq(self.param_env, ty, found).is_ok() {
                     // We only point at the first place where the found type was inferred.
-                    for (i, param_ty) in sig.inputs().skip_binder().iter().skip(1).enumerate() {
+                    for (param_ty, arg) in sig.inputs().skip_binder().iter().skip(1).zip(args) {
```

Comment:
Alternatively, we can just bail here if there is a argument arity mismatch. If there isn't the right # of arguments, there's probably no good reason to point at args as the source for a bad inference.

Reply:
Zipping here rather than indexing into `args` is nice regardless, we just might want to do this arity check regardless.

Reply:
Hm, in the example in the issue (#108042) the `.insert(..)` call has the wrong number of args but seems to still be the source of the inference? Or am I reading this wrong?

Current `rustc` output with the `.zip()` patch:

```
error[E0061]: this method takes 2 arguments but 1 argument was supplied
   --> p.rs:12:9
    |
12  |     map.insert(("a".to_string(), "a".to_string()));
    |         ^^^^^^------------------------------------ an argument is missing
    |
note: associated function defined here
   --> /home/mg/dev/rust/library/alloc/src/collections/btree/map.rs:955:12
    |
955 |     pub fn insert(&mut self, key: K, value: V) -> Option<V>
    |            ^^^^^^
help: provide the argument
    |
12  |     map.insert(("a".to_string(), "a".to_string()), /* value */);
    |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0308]: mismatched types
  --> p.rs:13:34
   |
13 |     let mut structdef = Struct { map };
   |                                  ^^^ expected `BTreeMap<String, String>`, found `BTreeMap<(String, String), _>`
   |
   = note: expected struct `BTreeMap<String, String>`
              found struct `BTreeMap<(String, String), _>`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
``

Reply:
I added (in 0be53945) a first stab at a UI test that shows what I mean, I think.

Reply:
Sure, but the inference is influenced in an incorrect way -- it's not super useful to point out this as the place where the inference is occurs when we have the other error E0061.